File: api/app/routers/onboarding.py
from __future__ import annotations
import logging, re
from typing import Optional
from fastapi import APIRouter, HTTPException, Header
from pydantic import BaseModel, field_validator
from app.services import onboarding_service

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/onboarding", tags=["onboarding"])

class RegisterRequest(BaseModel):
    handle: str
    email: Optional[str] = None
    hint_github: Optional[str] = None
    hint_wallet: Optional[str] = None

    @field_validator("handle")
    @classmethod
    def handle_format(cls, v: str) -> str:
        v = v.strip().lower()
        if not re.fullmatch(r"[a-z0-9_\-]{3,40}", v):
            raise ValueError("Handle must be 3-40 chars: a-z, 0-9, _ or -")
        return v

class RegisterResponse(BaseModel):
    contributor_id: str
    session_token: str
    trust_level: str
    handle: str
    created: bool
    roi_signals: dict

class SessionResponse(BaseModel):
    contributor_id: str
    handle: str
    trust_level: str
    linked_identities: int
    email: Optional[str]
    hint_github: Optional[str]
    hint_wallet: Optional[str]

class UpgradeRequest(BaseModel):
    contributor_id: str
    provider: str
    provider_id: str
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    metadata: Optional[dict] = None

@router.post("/register", response_model=RegisterResponse, summary="TOFU registration")
async def register(body: RegisterRequest) -> RegisterResponse:
    """Claim a unique handle, receive a session token. Trust-on-first-use (TOFU). Spec 168."""
    try:
        result = onboarding_service.register(
            handle=body.handle, email=body.email,
            hint_github=body.hint_github, hint_wallet=body.hint_wallet,
        )
    except ValueError as exc:
        if "handle_taken" in str(exc):
            raise HTTPException(status_code=409, detail="handle_taken")
        raise HTTPException(status_code=422, detail=str(exc))
    roi = onboarding_service.get_roi_signals()
    return RegisterResponse(
        contributor_id=result["contributor_id"], session_token=result["session_token"],
        trust_level=result["trust_level"], handle=result["handle"],
        created=result["created"], roi_signals=roi,
    )

@router.get("/session", response_model=SessionResponse, summary="Validate session token")
async def get_session(authorization: str = Header(...)) -> SessionResponse:
    """Return contributor profile for a Bearer session token. Returns 401 if invalid."""
    if not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="invalid_authorization_header")
    token = authorization.removeprefix("Bearer ").strip()
    profile = onboarding_service.resolve_session(token)
    if profile is None:
        raise HTTPException(status_code=401, detail="invalid_or_expired_token")
    return SessionResponse(**profile)

class ContributorListItem(BaseModel):
    contributor_id: str
    handle: str
    trust_level: str
    email: Optional[str]
    created_at: Optional[str]

@router.get("/contributors", response_model=list[ContributorListItem], summary="List registered contributors")
async def list_contributors(limit: int = 200) -> list[ContributorListItem]:
    """Return all registered contributors (newest first). Satisfies 'appear in contributor list' requirement."""
    rows = onboarding_service.list_contributors(limit=limit)
    return [ContributorListItem(**r) for r in rows]

@router.post("/upgrade", summary="Upgrade trust level via OAuth [stub 501]")
async def upgrade_oauth(body: UpgradeRequest) -> dict:
    """Stub: upgrade TOFU to verified via OAuth. Planned for Spec 169."""
    raise HTTPException(status_code=501, detail={
        "message": "OAuth upgrade not yet implemented. Planned for Spec 169.",
        "upgrade_path": {
            "github": "POST /api/identity/verify/github",
            "ethereum": "POST /api/identity/verify/ethereum",
        },
        "current_trust_level": "tofu",
        "spec_ref": "spec-168",
    })

@router.get("/roi", summary="Onboarding ROI signals")
async def roi_signals() -> dict:
    """Live ROI signals: handle_registrations, verified_ratio, avg_time_to_verify_days."""
    return onboarding_service.get_roi_signals()

File: cli/lib/commands/onboarding.mjs
/**
 * Onboarding commands: onboarding register, onboarding session, onboarding upgrade
 */

import { get, post } from "../api.mjs";

/** Register a new contributor (Trust-on-First-Use) */
export async function onboardingRegister(args) {
  let handle = null;
  let email = null;
  let hintGithub = null;
  let hintWallet = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--handle" && args[i + 1]) handle = args[++i];
    else if (args[i] === "--email" && args[i + 1]) email = args[++i];
    else if (args[i] === "--hint-github" && args[i + 1]) hintGithub = args[++i];
    else if (args[i] === "--hint-wallet" && args[i + 1]) hintWallet = args[++i];
  }

  if (!handle) {
    console.log("Usage: coh onboarding register --handle <name> [--email e] [--hint-github gh] [--hint-wallet addr]");
    return;
  }

  const body = { handle };
  if (email) body.email = email;
  if (hintGithub) body.hint_github = hintGithub;
  if (hintWallet) body.hint_wallet = hintWallet;

  const result = await post("/api/onboarding/register", body);
  if (!result) return;

  console.log(`\x1b[1m  REGISTERED\x1b[0m`);
  console.log(`  ${"─".repeat(50)}`);
  console.log(`  Contributor ID: ${result.contributor_id}`);
  console.log(`  Handle:         ${result.handle}`);
  console.log(`  Trust Level:    ${result.trust_level}`);
  console.log(`  Session Token:  ${result.session_token}`);
  console.log(`  Save this token securely for API access.`);
  if (result.roi_signals) {
    console.log(`  ROI Signals:    Registrations: ${result.roi_signals.handle_registrations}`);
  }
  console.log();
}

/** Get contributor profile from session token */
export async function onboardingSession(args) {
  let token = null;
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--token" && args[i + 1]) token = args[++i];
  }

  if (!token) {
    console.log("Usage: coh onboarding session --token <token>");
    return;
  }

  try {
    const { getHubUrl, getApiKey } = await import("../lib/config.mjs");
    const base = getHubUrl().replace(/\/$/, "");
    const headers = { "Content-Type": "application/json" };
    const key = getApiKey();
    if (key) headers["X-API-Key"] = key;
    headers["Authorization"] = `Bearer ${token}`;

    const res = await fetch(`${base}/api/onboarding/session`, {
      method: "GET",
      headers,
      signal: AbortSignal.timeout(12000),
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      console.error(`\x1b[2m  ${res.status}: ${err.detail || "request failed"}\x1b[0m`);
      return;
    }

    const data = await res.json();
    console.log(`\x1b[1m  SESSION\x1b[0m`);
    console.log(`  ${"─".repeat(50)}`);
    console.log(`  Contributor ID: ${data.contributor_id}`);
    console.log(`  Handle:         ${data.handle}`);
    console.log(`  Trust Level:    ${data.trust_level}`);
    console.log(`  Created:        ${new Date(data.created).toLocaleString()}`);
    if (data.email) console.log(`  Email:          ${data.email}`);
    if (data.hint_github) console.log(`  GitHub Hint:    ${data.hint_github}`);
    if (data.hint_wallet) console.log(`  Wallet Hint:    ${data.hint_wallet}`);
    console.log();
  } catch (err) {
    console.error(`\x1b[2m  network error: ${err.message}\x1b[0m`);
  }
}

/** Upgrade trust level to verified (stub) */
export async function onboardingUpgrade() {
  const result = await post("/api/onboarding/upgrade", {});
  if (!result) return;

  if (result.detail) {
    console.log(`  ${result.detail}`);
    console.log(`  Full OAuth flow coming in Spec 169.`);
    return;
  }

  console.log(`\x1b[1m  UPGRADED\x1b[0m`);
  console.log(`  Trust Level: ${result.trust_level}`);
  console.log();
}
